src/lib/define-dynamic-block.js

```diff
@@ -246,14 +246,45 @@ const defineDynamicBlock = (guiContext, ScratchBlocks, categoryInfo, staticBlock
                 }
             });
 
+            // Add a validator function to each field_dropdown which is responsible
+            // for keeping the mutation and the field value in sync. Emit a change
+            // event for the mutation so that the VM can also stay in sync with
+            // these changes.
+            const fieldValidator = argName => (function (selectedItem) {
+                // Disabling this lint rule since this function will get attached to the
+                // field_dropdown prototype appropriately.
+                /* eslint-disable no-invalid-this */
+                const oldMutation = ScratchBlocks.Xml.domToText(this.sourceBlock_.mutationToDom());
+                const currBlockInfo = JSON.parse(this.sourceBlock_.blockInfoText);
+                currBlockInfo.arguments[argName].selectedValue = selectedItem;
+                this.sourceBlock_.blockInfoText = JSON.stringify(currBlockInfo);
+                const newMutation = ScratchBlocks.Xml.domToText(this.sourceBlock_.mutationToDom());
+                ScratchBlocks.Events.fire(new ScratchBlocks.Events.BlockChange(this.sourceBlock_,
+                    'mutation', null, oldMutation, newMutation));
+                this.setValue(selectedItem);
+                return null;
+                /* eslint-enable no-invalid-this */
+            });
```

Comment:
It might be more readable to pull this function out: in particular, it might avoid confusion about `this`. Also, `defineDynamicBlock` is kind of too big already... :/

I'm imagining something like:
```js
const makeFieldValidator = function (argName) {
    return function (selectedItem) {
        const oldMutation = // ...
    };
};

const defineDynamicBlock = // ...
```

I suspect something like this might make the `eslint-disable no-invalid-this` unnecessary, too, but I'm not certain.

Reply:
Hmmm I agree that `defineDynamicBlock` is getting much too bloated. When pulling this function out though, we need to pass in ScratchBlocks as an argument, and unfortunately the `this` issue does not go away (still need the eslint-disables).

Reply:
Moved the `makeFieldValidator` functionality into its own function as you suggested. I think even with passing in `ScratchBlocks` as an argument, it helps break up the complexity of `domToMutation` a little bit.